[dtensor] implement scatter op with simple replication

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126713
* #126712

as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

wanchaol committed:
[dtensor] implement scatter op

as titled, implement torch.scatter op

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op"


as titled, implement torch.scatter op

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op"


as titled, implement torch.scatter op

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op"


as titled, implement torch.scatter op

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op"


as titled, implement torch.scatter op

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op"


as titled, implement torch.scatter op

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

[ghstack-poisoned]

tianyu-l commented on test/distributed/_tensor/test_tensor_ops.py:
```diff
@@ -390,6 +390,36 @@ def test_new_empty_strided(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        comm_mode = CommDebugMode()
+
+        # case 1 all replicate: input replicated, index replicated, output replicated
```
```suggestion
        # case 1 all replicate: input replicated, index replicated, src replicated, output replicated
```

tianyu-l commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -359,6 +360,28 @@ def replica_only_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType
 
 
+@register_op_strategy(
+    [aten.scatter_.value, aten.scatter.value, aten.scatter_.src, aten.scatter.src],
```
This replicate strategy should also work for other reduce variants, e.g. `scatter.reduce`, `scatter_.reduce`, etc.

wanchaol replied:
I think these ops are being deprecated, so I'll leave them as separate PRs

tianyu-l commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -359,6 +360,28 @@ def replica_only_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType
+
+    # placement list stores placements of [output, input, index, src]
+    # first we always have replicate all for inputs and output
+    all_replicate: List[Placement] = [Replicate()] * 4
```
For `aten.scatter_.value` and `aten.scatter.value`, `src` is a value rather than a tensor, although I think the code here should still work in those cases. Can we comment here and add a test case where `src` is a float number?

tianyu-l reviewed (approved):
Looks good to me. Had some inline comments.
The sharded OpStrategy for `scatter` might be somewhat different than `gather`, especially considering there are sub-variants with different `reduce` type.

My feelings is that
1) We at least need other types of `Partial` support (not just `_MaskPartial`).
2) It might makes sense to implement sharded strategy later until we see a concrete use case. If you look at the sharded strategy for `gather`, they are also special and motivated by clear use case when we were working on Loss Parallel (although finally it was not used directly in Loss Parallel).

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scatter op with simple replication"


as titled, implement torch.scatter op with simple replications strategy,
need to follow up and see if we could actually support any sharding
pattern

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "forever queueing macos job"
